draft-ietf-tls-esni.md

```diff
@@ -376,75 +402,63 @@ multiple "outer_extension" extensions with the same extension code point.
 
 # Client Behavior {#client-behavior}
 
-## Sending an encrypted ClientHello {#send-esni}
+## Sending an encrypted ClientHello {#send-echo}
 
-In order to send an encrypted SNI, the client MUST first generate its
+In order to send an encrypted ClientHello, the client MUST first generate its
 ClientHelloInner value. In addition to the normal values, ClientHelloInner
 MUST also contain:
 
- - an "esni_nonce" extension
- - a TLS padding {{!RFC7685}}. This SHOULD contain X bytes of padding
-   where X + the actual server name is equal to ESNIConfig.maximum_name_length
-
-Then, the client MUST select one of the server ESNIKeyShareEntry
-values and generate an (EC)DHE share in the matching group. This share
-will then be sent to the server in the "encrypted_client_hello"
-extension and used to derive the SNI encryption key. It does not
-affect the (EC)DHE shared secret used in the TLS key schedule. The
-client MUST also select an appropriate cipher suite from the list of
-suites offered by the server. If the client is unable to select an
-appropriate group or suite it SHOULD ignore that ESNIConfig value and
-MAY attempt to use another value provided by the server. The client
-MUST NOT send encrypted SNI using groups or cipher suites not
-advertised by the server.
-
-When offering an encrypted SNI, the client MUST NOT offer to resume any non-ESNI
+ - an "echo_nonce" extension
+ - TLS padding {{!RFC7685}}. This SHOULD contain X bytes of padding
+   where X + the length of the non-padding bytes in ClientHelloInner
+   is equal to ECHOConfig.maximum_message_length.
```

Comment:
Editorial nitpick: it seems this section (assembling ClientHelloInner based on an ECHOConfig) ought to come after the logic below (determining which ECHOConfig to use, if any).

Reply:
Agreed -- I'll shuffle text around.

Reply:
👍